### PR TITLE
annotation of pod/ns/dep with an invalid EPG will be shown as a fault

### PR DESCRIPTION
**Pull request by Bhavanaashok33**
Fault will be raised against ns/pod/deployment if the epg doesnt exist on APIC
Added unit tests 
Tested the changes on fabric 
Signed-off-by: bashokba <bashokba@cisco.com>

### CODE REVIEW

**Comment by tomflynn** [comment]
Can you enhance the apic_types_test.go to cover the new types you've added.
Can you also update the UTs so it tests the fault creation? coveralls shows those lines aren't being covered by UTs currently

**Comment by Bhavanaashok33** [comment]
I have addressed all the review comments. wrt fault creation test on APIC, I have simulated a mock APIC(local http server) that accepts the REST API and partially mock's the APIC.  Currently Testing is it process. please don't review the code yet..

**Comment by Bhavanaashok33** [comment]
Please go ahead and review the code! 
Enhanced the UTs covering fault creating on mock APIC.

File: pkg/controller/annotation_test.go [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+	"\"app-profile\": \"test\", \"name\": \"test-qp\"}"
+
+var vrfEpgDn string = "uni/testps/test/test-eg"
+var cachedVRFDn []string
+
+type loginSucc struct {
+	cert bool
+}
+
+type socketHandler struct {
```
Comment (jayaramsatya) [inline pkg/controller/annotation_test.go:65]: you can reference this structure apicapi, please check where ever you can.

File: pkg/controller/pods.go [hunk 2]
```diff
@@ -92,29 +101,39 @@ func (cont *AciController) checkIfEpgExistPod(pod *v1.Pod) {
...
-	notExist, egval := cont.checkVrfCache(epGroup, "pod[EpgAnnotation]")
-
-	if notExist && egval.Name != "" {
-		desc := "Pod Annotation failed for the pod " + pod.Name + "(Namespace " +
-			pod.Namespace + ", Reason being: " + "Cannot resolve the EPG " + egval.Name +
-			" for the tenant " + egval.Tenant + " and app-profile " + egval.AppProfile
-		faultCode := strconv.Itoa(10)
-		severity := "major"
-		cont.createFaultInst(key, desc, faultCode, severity)
+	if len(pod.ObjectMeta.Annotations[metadata.EgAnnotation]) > 0 {
```
Comment (jayaramsatya) [inline pkg/controller/pods.go:104]: Please change this line 
epGroup, ok  := pod.ObjectMeta.Annotations[metadata.EgAnnotation]
if ok {
  cont.handleEpgAnnotationUpdate(key, 10, major, pod.Name, epGroup)
}

File: pkg/controller/subnetcache.go [hunk 3]
```diff
@@ -132,7 +130,7 @@ func (cont *AciController) BuildSubnetDnCache(dn string, aciVrfDn string) {
 	}
 	sort.Strings(vrfEpgDns)
 	cont.log.Debug("aciVrfEpgDns: ", vrfEpgDns)
-	cachedVRFDns = vrfEpgDns
+	cont.cachedVRFDns = vrfEpgDns
```
Comment (jayaramsatya) [inline pkg/controller/subnetcache.go:133]: can you check if you need to lock this?

File: pkg/controller/pods.go [hunk 4]
```diff
@@ -31,6 +32,15 @@ import (
 	"k8s.io/client-go/tools/cache"
 )
 
+const (
+	cleared  = iota
```
Comment (jayaramsatya) [inline pkg/controller/pods.go:36]: please give the name to the severity

File: pkg/controller/namespaces.go [hunk 5]
```diff
@@ -126,24 +125,16 @@ func (cont *AciController) namespaceDeleted(obj interface{}) {
...
-	notExist, egval := cont.checkVrfCache(epGroup, "namespace[EpgAnnotation]")
-
-	if notExist && egval.Name != "" {
-		desc := "Annotation failed for the Namespace " + namespaceobj.Name +
-			", Reason being: " + " Cannot resolve the EPG " + egval.Name +
-			" for the tenant " + egval.Tenant + " and app-profile " + egval.AppProfile
-		faultCode := strconv.Itoa(11)
-		severity := "critical"
-		cont.createFaultInst(nskey, desc, faultCode, severity)
+	if len(namespaceobj.ObjectMeta.Annotations[metadata.EgAnnotation]) > 0 {
```
Comment (jayaramsatya) [inline pkg/controller/namespaces.go:134]: change it same as above  comment

**Review by jayaramsatya — APPROVED** [review]
LGTM

**Review by tomflynn — APPROVED** [review]

**Pull request merged**